File: src/medium/_475_供暖器.java
package medium;

import java.util.Arrays;

public class _475_供暖器 {
    public int findRadius(int[] houses, int[] heaters) {
        // 排序
        Arrays.sort(houses);
        Arrays.sort(heaters);
        int ans = 0;
        for (int i = 0, j = 0; i < houses.length; i++) {
            //拿到 当前的节点 和 指定节点的供暖器 的距离
            int curDistance = Math.abs(houses[i] - heaters[j]);
            // 如果 下一个供暖器离该节点更近一些，挪动到下一个节点
            while (j < heaters.length - 1 && Math.abs(houses[i] - heaters[j]) >= Math.abs(houses[i] - heaters[j + 1])) {
                j++;
                curDistance = Math.min(curDistance, Math.abs(houses[i] - heaters[j]));
            }
            ans = Math.max(ans, curDistance);
        }
        return ans;
    }
}

File: src/go/medium/findRadius/475.供暖器.go
package findRadius

import (
	"sort"
)

//1. 对于每个房屋，要么用前面的暖气，要么用后面的，二者取近的，得到距离；
//2. 对于所有的房屋，选择最大的上述距离。
func findRadius(houses []int, heaters []int) int {
	sort.Ints(houses)
	sort.Ints(heaters)
	ans := 0
	j := 0
	for i := 0; i < len(houses); i++ {
		distance := abs(houses[i] - heaters[j])
		for j < len(heaters)-1 && abs(houses[i]-heaters[j+1]) <= distance {
			j++
			distance = abs(houses[i] - heaters[j])
		}
		if ans < distance {
			ans = distance
		}
	}
	return ans
}
func abs(b int) int {
	if b < 0 {
		return -b
	}
	return b
}
